File: csharp/robot-name/RobotName.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class Robot
{
    public string Name { get; private set; }

    public Robot() => Name = RobotNameGenerator.GenerateUniqueName();

    public void Reset() => Name = RobotNameGenerator.GenerateUniqueName();
}

internal static class RobotNameGenerator
{
    private const string Alpha = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
    private const string Numeric = "0123456789";

    private static readonly string[] Recipe = { Alpha, Alpha, Numeric, Numeric, Numeric };
    private static readonly int NumberOfUniqueNames = Recipe.Aggregate(1, (acc, seq) => acc * seq.Length);

    private static readonly ISet<string> UsedNames = new HashSet<string>();

    public static string GenerateUniqueName()
    {
        if (UsedNames.Count == NumberOfUniqueNames)
        {
            throw new Exception("Number of unique names exhausted");
        }

        string name;
        do
        {
            name = GenerateName();
        } while (!UsedNames.Add(name));

        return name;
    }

    private static string GenerateName()
    {
        var rand = new Random();
        char RandomCharacter(string sequence) => sequence[rand.Next(sequence.Length - 1)];

        return string.Join("", Recipe.Select(RandomCharacter));
    }
}

File: kotlin/robot-name/src/main/kotlin/RobotName.kt
class Robot {
    companion object {
        private val namesInUse = mutableSetOf<String>()
        private val alpha = ('A'..'Z')
        private val numeric = ('0'..'9')
        private val nameRecipe = listOf(alpha, alpha, numeric, numeric, numeric)
        private val maxNumberOfNames = nameRecipe.fold(1) { n, r -> n * r.toList().size }
    }

    lateinit var name: String
        private set

    init {
        reset()
    }

    fun reset() {
        require(namesInUse.size < maxNumberOfNames) { "Robot names exhausted" }

        name = generateSequence(::generateName).first { namesInUse.add(it) }
    }

    private fun generateName() = nameRecipe.map { it.random() }.joinToString("")
}
